package_modules/apt_get

```diff
@@ -0,0 +1,259 @@
+        #           V              V                    V                        V                 V
+        #   Inst php5-cli [5.3.10-1ubuntu3.17] (5.3.10-1ubuntu3.18 Ubuntu:12.04/precise-updates [amd64]) []
+        #
+        # Note architecture included in the name on this one:
+        #   Inst php5-cli:i386 [5.3.10-1ubuntu3.17] (5.3.10-1ubuntu3.18 Ubuntu:12.04/precise-updates [i386]) []
+        #
+        # Note multiple repositories in this one:
+        #   Inst linux-libc-dev [2.6.32-48squeeze4] (2.6.32-48squeeze6 Debian:6.0.10/oldstable, Debian-Security:6.0/oldoldstable [amd64])
+        #
+        match = re.match("^Inst\s+([^\s:]+)(?::\S+)?\s+\[[^]\s]+\]\s+\((\S+)\s+[^,\s]+(?:,\s+[^,\s]+)*\s+\[([^]\s]+)\]\).*", line)
+        if match != None:
+            sys.stdout.write("Name=" + match.group(1) + "\n")
+            sys.stdout.write("Version=" + match.group(2) + "\n")
+            sys.stdout.write("Architecture=" + match.group(3) + "\n")
+
+    return 0
+
+
+def one_package_argument(name, arch, version, is_apt_install):
+    args = []
+    archs = []
+    if arch != "":
+        archs.append(arch)
+    elif is_apt_install:
+        # Here we have no specified architecture and we are installing.
+        # If we have existing versions, operate on those, instead of the
+        # platform default.
+        process = subprocess.Popen([dpkg_query_cmd, "--showformat", "${Architecture}=${Status}\n",
+                                    "-W", name + ":*"], stdout=subprocess.PIPE)
+        for line in process.stdout:
+            # The space before "installed" is important, because it can be "not-installed".
+            match = re.match("^([^=]*)=.* installed.*", line)
+            if match:
+                archs.append(match.group(1))
+
+    version_suffix = ""
+    if version != "" and is_apt_install:
+        version_suffix = "=" + version
+
+    if len(archs) > 0:
+        for arch in archs:
+            args.append(name + ":" + arch + version_suffix)
+    elif is_apt_install:
+        args.append(name + version_suffix)
+    else:
+        # Delete all architectures if none is specified.
+        args.append(name + ":*" + version_suffix)
+
+    return args
+
+
+def package_arguments_builder(is_apt_install):
+    name = ""
+    version = ""
+    arch = ""
+    args = []
+    for line in sys.stdin:
+        result = re.match("^Name=(.*)", line)
+        if (result):
+            if (name != ""):
+                # Each new "Name=" triggers a new entry.
+                args += one_package_argument(name, arch, version, is_apt_install)
+
+                version = ""
+                arch = ""
+
+            name = result.group(1)
+            continue
+
+        result = re.match("^Version=(.*)", line)
+        if (result):
+            version = result.group(1)
+            continue
+
+        result = re.match("^Architecture=(.*)", line)
+        if (result):
+            arch = result.group(1)
+            continue
+
```

Comment:
```
for line in sys.stdin:
    if line.startswith('Name='):
         if name:
            args += one_package_argument(name, arch, version, is_apt_install)
            version = arch = ''
         name = line[5:]
elif line.startswith('Version='): version = line[8:]
elif line.startswith('Architecture='): arch = line[13:]
```
Because you don't need to do the matching outside the flow of the `if-elif` chain, you save yourself a stack of bureaucracy and the need to continue.